orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+      property_file_path, true,
+      [&property_protos](int fd) {
+        google::protobuf::io::FileInputStream input{fd};
+        ORT_RETURN_IF_ERROR(ReadProtoMessageSequence(property_protos, input));
+        return Status::OK();
+      });
+
+  if (!file_read_status.IsOK()) {
+    LOGS_DEFAULT(WARNING) << ToUTF8String(property_file_path) << " custom property file read failure, skip it.";
+    return Status::OK();
+  }
+
+  std::vector<std::shared_ptr<CheckpointProperty>>& custom_properties = states.custom_properties;
+  for (auto& property_proto : property_protos) {
+    const std::string& tensor_name = property_proto.name();
+    auto data_type = property_proto.data_type();
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto::FLOAT: {
+        const std::vector<float>& flt_parsed = ONNX_NAMESPACE::ParseData<float>(&property_proto);
+        ORT_ENFORCE(flt_parsed.size() == static_cast<size_t>(1), "only support scalar float properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<float>>(
+            tensor_name, flt_parsed.at(0)));
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto::STRING: {
+        const std::vector<std::string>& str_parsed = ParseStringData(&property_proto);
+        ORT_ENFORCE(str_parsed.size() == static_cast<size_t>(1), "only support scalar string properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<std::string>>(
+            tensor_name, str_parsed.at(0)));
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto::INT64: {
+        const std::vector<int64_t>& int_parsed = ONNX_NAMESPACE::ParseData<int64_t>(&property_proto);
+        ORT_ENFORCE(int_parsed.size() == static_cast<size_t>(1), "only support scalar int64_t properties.");
+        custom_properties.emplace_back(std::make_shared<TypedCheckpointProperty<int64_t>>(
+            tensor_name, int_parsed.at(0)));
+        break;
+      }
+      default:
+        ORT_THROW("Unsupported input data type of ", data_type);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtLoadModuleStatesInternal(
+    const PathString& parameter_folder_path, ModuleCheckpointStates& module_states) {
+  // Parameter parsing.
+  const PathString module_state_file_path = GetTensorProtoFilePath(parameter_folder_path, k_param_root_prefix);
+  auto& named_parameters = module_states.named_parameters;
+  std::vector<ONNX_NAMESPACE::TensorProto> param_tensor_protos{};
+  auto file_read_status = WithOpenFile(
+      module_state_file_path, true,
+      [&param_tensor_protos](int fd) {
+        google::protobuf::io::FileInputStream input{fd};
+        ORT_RETURN_IF_ERROR(ReadProtoMessageSequence(param_tensor_protos, input));
+        return Status::OK();
+      });
+
+  if (!file_read_status.IsOK()) {
+    LOGS_DEFAULT(WARNING) << ToUTF8String(module_state_file_path) << " module state file read failure, skip it.";
```

Comment:
Shouldn't this be an error

Reply:
My thinking is, either part (among module, optimizer, custom properties) of the checkpoint can be optional, they are independent.

Reply:
Fixed in the latest commit. Instead of just load the file without detecting whether the folder has the file, I added the logic firstly probe the directory found all candidate files; if none, then skip; if there is, load it with error check.